### PR TITLE
parser: support sql_mode "PIPES_AS_CONCAT"

### PR DESCRIPTION
**Pull request by spongedu**
support sql mode "PIPES_AS_CONCAT".

### CODE REVIEW

File: parser/parser.y [hunk 1]
```diff
@@ -430,6 +430,8 @@ import (
 	paramMarker	"?"
 	rsh		">>"
 
+%token<ident> or2
```
Comment (zz-jason) [inline parser/parser.y]: what does `or2` means ?

Comment (spongedu) [reply parser/parser.y]: represent `||` when `PIPES_AS_CONCQT` is not set. I think it's better to distinguish `||`  under different sql_modes, and that's how MySQL do

Comment (zz-jason) [reply parser/parser.y]: @winkyao PTAL

File: parser/parser.y [hunk 2]
```diff
@@ -5456,6 +5463,9 @@ DateAndTimeType:
 		x := types.NewFieldType(mysql.TypeTimestamp)
 		x.Flen = mysql.MaxDatetimeWidthNoFsp
 		x.Decimal = $2.(int)
+		if x.Decimal == types.UnspecifiedLength{
```
Comment (coocood) [inline parser/parser.y]: Why change this?

**Comment by coocood** [comment]
How about rename token `||` to `pipes`? 
It would be more clear.
`pipes` is the token name, it can be used as "or" or "concat".

File: parser/parser.y (hunk 2 above)
Comment (spongedu) [reply parser/parser.y]: it's changed by mistake. I'll revert it.

Comment (spongedu) [reply parser/parser.y]: @coocood  it's change by mistake. I'll revert it.

**Comment by spongedu** [comment]
@coocood  I think the token name issue can be considered on both sides:
1.  rename `||` to `pipes` is clearer, definitely.
2. current form is more similar to what mysql do. Maybe it's easier to follow later changes in the future. 
What do you think?

**Comment by coocood** [comment]
I think we can afford this change, many of our token names are not exactly the same as MySQL.

And, don't forget to align `parser.y` by tab.

**Comment by spongedu** [comment]
@coocood ok. I'll fix this soon

**Comment by winkyao** [comment]
Please fix conflict. @spongedu

File: parser/parser.y [hunk 3]
```diff
@@ -434,6 +434,8 @@ import (
 	paramMarker	"?"
 	rsh		">>"
 
+%token<ident> pipesAsOr
```
Comment (winkyao) [inline parser/parser.y]: move this line to line 327

**Comment by spongedu** [comment]
Done @winkyao . PTAL

**Comment by coocood** [comment]
LGTM

**Pull request merged**